libp2p/Session.cpp

```diff
@@ -459,3 +522,89 @@ bool Session::checkRead(std::size_t _expected, boost::system::error_code _ec, st
 
 	return true;
 }
+
+void Session::doReadFrames()
+{
+	if (m_dropped)
+		return; // ignore packets received while waiting to disconnect
+
+	auto self(shared_from_this());
+	m_data.resize(h256::size);
+	ba::async_read(m_socket->ref(), boost::asio::buffer(m_data, h256::size), [this, self](boost::system::error_code ec, std::size_t length)
+	{
+		ThreadContext tc(info().id.abridged());
+		ThreadContext tc2(info().clientVersion);
+		if (!checkRead(h256::size, ec, length))
+			return;
+
+		DEV_GUARDED(x_writeQueue)
+		{
+			if (!m_io->authAndDecryptHeader(bytesRef(m_data.data(), length)))
+			{
+				clog(NetWarn) << "header decrypt failed";
+				drop(BadProtocol); // todo: better error
+				return;
+			}
+		}
+
+		bytesConstRef rawHeader(m_data.data(), length);
+		try
+		{
+			RLPXFrameInfo tmpHeader(rawHeader);
+		}
+		catch (std::exception const& _e)
+		{
+			clog(NetWarn) << "Exception decoding frame header RLP:" << _e.what() << bytesConstRef(m_data.data(), h128::size).cropped(3);
+			drop(BadProtocol);
+			return;
+		}
+
+		RLPXFrameInfo header(rawHeader);
+		auto tlen = header.length + header.padding + h128::size; // padded frame and mac
+		m_data.resize(tlen);
+		ba::async_read(m_socket->ref(), boost::asio::buffer(m_data, tlen), [this, self, tlen, header](boost::system::error_code ec, std::size_t length)
+		{
+			ThreadContext tc(info().id.abridged());
+			ThreadContext tc2(info().clientVersion);
+			if (!checkRead(tlen, ec, length))
+				return;
+
+			bytesRef frame(m_data.data(), tlen);
+			vector<RLPXPacket> px;
+			DEV_GUARDED(x_writeQueue)
+			{
+				auto v = getFraming(header.protocolId)->reader.demux(*m_io, header, frame);
+				px.swap(v);
+			}
+
+			for (RLPXPacket& p: px)
+			{
+				PacketType packetType = (PacketType)RLP(p.type()).toInt<unsigned>(RLP::AllowNonCanon);
+				bool ok = readPacket(header.protocolId, packetType, RLP(p.data()));
+#if ETH_DEBUG
+				if (!ok)
+					clog(NetWarn) << "Couldn't interpret packet." << RLP(p.data());
+#endif
+				ok = true;
+			}
+			doReadFrames();
+		});
+	});
+}
+
+std::shared_ptr<Session::Framing> Session::getFraming(uint16_t _protocolID)
+{
+	if (m_framing.find(_protocolID) == m_framing.end())
+	{
+		std::shared_ptr<Session::Framing> p(new Session::Framing(_protocolID));
```

Comment:
This method shouldn't automatically add/generate Framing objects. Framing objects should be created when the capability is registered or destroyed when it is deregistered.